### DISCUSSION TITLE
Dependency injection?

### DISCUSSION THREAD

**Discussion by DenuxPlays** [actix/actix-web#3270] [Q&A] [answered]
Hello,
I've wondered if there is a way to implement "dependency injection" of some kind.
My basic goal is to make this workl:
```rs
#[get("/user/{user_id}")]
pub async fn get_one(user: User) -> impl Responder {
    // Some more operations
    retrun HttpResponse::Ok().json(user);
}
```
The things is that user needs to have access to the user_id and the identity (from actixcidentity crate).
I found the [FromRequest](https://docs.rs/actix-web/latest/actix_web/trait.FromRequest.html)  trait but not sure if I can make this work with that or if there is a simpler way.
Of course I could implement that logic  for every endpoint but this would result in a lot of code duplication even if I create an extra function that returns the user or an error.

Is there any way to make this work?

**Comment by DenuxPlays** [comment]
Edit:
The FromRequest trait is not async so as I use sea-orm (an async ORM) I cannot utilize the FromRequest trait.

**Reply by robjtede** [reply, marked as answer]
The FromRequest trait support async operations, see docs: https://docs.rs/actix-web/latest/actix_web/trait.FromRequest.html#associatedtype.Future

I'd suggest that creating an extractor (impl FromRequest) is the right way of achieving your goal here.

**Reply by DenuxPlays** [reply]
> The FromRequest trait support async operations, see docs: https://docs.rs/actix-web/latest/actix_web/trait.FromRequest.html#associatedtype.Future
> 
> 
> 
> I'd suggest that creating an extractor (impl FromRequest) is the right way of achieving your goal here.

Ah okay 
Gonna try this sounds promising 
Thank you 🙏
